plugins/dig-now.cpp

```diff
@@ -26,12 +29,102 @@
 #include <df/world.h>
 #include <df/world_site.h>
 
+#include <cinttypes>
+#include <unordered_set>
+#include <unordered_map>
+
 DFHACK_PLUGIN("dig-now");
 REQUIRE_GLOBAL(plotinfo);
 REQUIRE_GLOBAL(world);
 
+// Debugging
+namespace DFHack {
+    DBG_DECLARE(dignow, general, DebugCategory::LINFO);
+    DBG_DECLARE(dignow, channels, DebugCategory::LINFO);
+}
+
+#define COORD "%" PRIi16 " %" PRIi16 " %" PRIi16
+#define COORDARGS(id) id.x, id.y, id.z
+
+// todo: integrate logging for debugging the layered channel problem
+
 using namespace DFHack;
 
+struct designation{
+    df::coord pos;
+    df::tile_designation type;
+    df::tile_occupancy occupancy;
+    designation(const df::coord &c, const df::tile_designation &td, const df::tile_occupancy &to) : pos(c), type(td), occupancy(to) {}
+
+    bool operator==(const designation &rhs) const {
+        return pos == rhs.pos;
+    }
+
+    bool operator!=(const designation &rhs) const {
+        return !(rhs == *this);
+    }
+};
+
+namespace std {
+    template<>
+    struct hash<designation> {
+        std::size_t operator()(const designation &c) const {
+            std::hash<df::coord> hash_coord;
+            return hash_coord(c.pos);
+        }
+    };
+}
+
+class DigJobs {
+private:
+    std::unordered_map<df::coord, df::tile_dig_designation> designations;
+    std::unordered_map<df::coord, df::job*> jobs;
+public:
+    void load() {
+        designations.clear();
+        df::job_list_link* node = df::global::world->jobs.list.next;
+        while (node) {
+            df::job* job = node->item;
+            jobs.emplace(job->pos, job);
+            node = node->next;
+            switch (job->job_type){
+                case df::enums::job_type::Dig:
+                    designations.emplace(job->pos, df::tile_dig_designation::Default);
+                    break;
+                case df::enums::job_type::DigChannel:
+                    designations.emplace(job->pos, df::tile_dig_designation::Channel);
+                    break;
+                case df::enums::job_type::CarveRamp:
+                    designations.emplace(job->pos, df::tile_dig_designation::Ramp);
+                    break;
+                case df::enums::job_type::CarveUpwardStaircase:
```

Comment:
This needs to be revised to put together and store `designation` objects, and the switch expanded to include the other designations dig-now cares about.

Reply:
here's some code I put together while I was investigating this. It's mostly duplicative of what you already have, but it might help a little:
```cpp
static void get_dig_jobs(color_ostream &out,
        std::unordered_map<df::coord, df::job *> &jobs) {
    for (df::job_list_link *link = &world->jobs.list; link; link = link->next) {
        df::job *job = link->item;

        if (!job)
            continue;

        if (job->job_type != df::enums::job_type::Dig &&
                job->job_type != df::enums::job_type::CarveUpwardStaircase &&
                job->job_type != df::enums::job_type::CarveDownwardStaircase &&
                job->job_type != df::enums::job_type::CarveUpDownStaircase &&
                job->job_type != df::enums::job_type::CarveRamp &&
                job->job_type != df::enums::job_type::DigChannel)
            continue;

        jobs.emplace(job->pos, job);
    }
}

df::tile_dig_designation get_dig_job_designation(df::job *job) {
    switch (job->job_type) {
        case df::enums::job_type::Dig:
            return df::tile_dig_designation::Default;
        case df::enums::job_type::CarveUpwardStaircase:
            return df::tile_dig_designation::UpStair;
        case df::enums::job_type::CarveDownwardStaircase:
            return df::tile_dig_designation::DownStair;
        case df::enums::job_type::CarveUpDownStaircase:
            return df::tile_dig_designation::UpDownStair;
        case df::enums::job_type::CarveRamp:
            return df::tile_dig_designation::Ramp;
        case df::enums::job_type::DigChannel:
            return df::tile_dig_designation::Channel;
        default:
           return df::tile_dig_designation::No;
    }
}
```

Reply:
```diff
@@ -597,6 +637,9 @@ static void do_dig(color_ostream &out, std::vector<df::coord> &dug_coords,
 
     rng.init();
 
+    std::unordered_map<df::coord, df::job *> jobs;
+    get_dig_jobs(out, jobs);
+
     // go down levels instead of up so stacked ramps behave as expected
     for (int16_t z = options.end.z; z >= options.start.z; --z) {
         for (int16_t y = options.start.y; y <= options.end.y; ++y) {
@@ -608,8 +651,12 @@ static void do_dig(color_ostream &out, std::vector<df::coord> &dug_coords,
 
                 df::coord pos(x, y, z);
                 df::tile_designation td = map.designationAt(pos);
+                df::tile_dig_designation tdd = td.bits.dig;
                 df::tile_occupancy to = map.occupancyAt(pos);
-                if (td.bits.dig != df::tile_dig_designation::No &&
+                if (jobs.count(pos)) {
+                    tdd = get_dig_job_designation(jobs[pos]);
+                }
+                if (tdd != df::tile_dig_designation::No &&
                         !to.bits.dig_marked) {
                     std::vector<dug_tile_info> dug_tiles;
                     if (dig_tile(out, map, pos, td.bits.dig, dug_tiles)) {
```

Reply:
but of course that doesn't handle the smooth and track cases.

Reply:
Finding the smooth/track info in the job* is my next task. Iirc there is an issue about smoothing designations that may get linked as well.